### ISSUE TITLE
Bug on KabyLake-R

### ISSUE THREAD

**Issue by trihutama** [psyq321/PowerMonkey#10]
I successfully boot powermonkey.efi to undervolt my cpu with parameter:
Core: -65mV
Ring (Cache): -65mV

my cpu is i5 8250U , it's a laptop CPU (KabyLake-R)
Secure Boot: Disable
Virtual: Enable

but there also a bug where it overrides other settings like iGPU/System agent.
![image](https://user-images.githubusercontent.com/24309814/157015825-722a1da9-6f82-48ed-a6e7-3bc697584c8b.png)

This is my configuration :

`  
   /////////////////////////////////////
    /// Voltage / Frequency Overrides ///
    /////////////////////////////////////

    //
    // Select which domains are to be programmed
    
    pk->Program_VF_Overrides[IACORE] =    1;    // Enable programming of VF
                                                // Overrides for IA Cores

    //
    // Hybrid architectures (Alder Lake+) - E-Cores

    pk->Program_VF_Overrides[ECORE] = 0;        // Enable programming of VF
                                                // Overrides for E-Cores

    pk->Program_VF_Overrides[RING] =  1;        // Enable programming of VF
                                                // Overrides for Ring / Cache

    pk->Program_VF_Overrides[UNCORE] =    0;    // Enable programming of VF
                                                // Overrides for Uncore (SA)

    pk->Program_VF_Overrides[GTSLICE] =   1;    // Enable programming of VF
                                                // Overrides for GT Slice

    pk->Program_VF_Overrides[GTUNSLICE] = 1;    // Enable programming of VF
                                                // Overrides for GT Unslice

    // NOTE: some domains are sharing the same voltage plane! Check yours!
    // 
    // E.g.: for SKL, CFL, CML and RKL: IACORE (CPU cores) and RING (cache) 
    // share a common VR. If you don't program both linked domains to exactly 
    // the same voltage CPU's pcode will use higher voltage and apply it to both
    // domains - but without adjusting values submitted by the user so it 
    // appears everything went as user requested! Some might believe 
    // they won the 'chip lottery' seeing their CPU seemingly undervolt to 
    // -250 mV or so, while in reality pcode is doing exactly nothing!
    // Don't be that guy (or girl)!

    ///
    /// LEGACY V/F OVERRIDES FOR DOMAINS: CORE & RING (Cache)
    /// Note: Shared VR on test CML-H CPU = Same Voltages Applied
    /// 
    /// For CPU SKUs allowing V/F point adjustment you can adjust individual 
    /// points on the V/F curve instead (see below)
    /// 
    /// NOTE: Legacy V/F overrides >will not apply< if you chose to program
    /// individual V/F points. 
    
    ///////////////////////////////
    // Legacy (entire VF curve)  //
    ///////////////////////////////

    pk->planes[IACORE].VoltMode = 
      pk->planes[RING].VoltMode = V_IPOLATIVE;      // V_IPOLATIVE = Interpolate
                                                    // V_OVERRIDE =  Override
    pk->planes[IACORE].TargetVolts = 
      pk->planes[RING].TargetVolts = 0;             // in mV (absolute)

    pk->planes[IACORE].OffsetVolts = 
    pk->planes[RING].OffsetVolts = -65;// -50;        // in mV 
                                                    // (negative = undervolt)
    
    ///
    /// V/F OVERRIDES FOR DOMAIN: E-Core (ADL & Co. only)
    ///
    ///
    /// V/F OVERRIDES FOR DOMAIN: UNCORE (SA)
    ///

    // Add your adjustments here if needed    
	pk->planes[UNCORE].VoltMode = V_IPOLATIVE;  // V_IPOLATIVE = Interpolate
                                                // V_OVERRIDE =  Override
    pk->planes[UNCORE].TargetVolts = 0;         // in mV (absolute)
    pk->planes[UNCORE].OffsetVolts = 0;// -35;  // in mV (negative = undervolt)
    ///
    /// V/F OVERRIDES FOR DOMAIN: GT SLICE
    ///
	pk->planes[GTSLICE].VoltMode = V_IPOLATIVE;  // V_IPOLATIVE = Interpolate
                                                // V_OVERRIDE =  Override
    pk->planes[GTSLICE].TargetVolts = 0;         // in mV (absolute)
    pk->planes[GTSLICE].OffsetVolts = -55;// -35;  // in mV (negative = undervolt)
    // Add your adjustments here if needed

    ///
    /// V/F OVERRIDES FOR DOMAIN: GT UNSLICE
    ///
    pk->planes[GTUNSLICE].VoltMode = V_IPOLATIVE;  // V_IPOLATIVE = Interpolate
                                                // V_OVERRIDE =  Override
    pk->planes[GTUNSLICE].TargetVolts = 0;         // in mV (absolute)
    pk->planes[GTUNSLICE].OffsetVolts = -55;// -35;  // in mV (negative = undervolt)
    // Add your adjustments here if needed
`
i would like tracing the problem but i don't know where to start.

**Comment by psyq321** [comment]
Check the code above, for example:

pk->planes[IACORE].OffsetVolts = 
pk->planes[RING].OffsetVolts = -65;// -50;        // in mV 
                                                // (negative = undervolt)

This means we are setting the same value for both IACORE and RING domains - if you want this different, you should split (but please check if that is supported by your CPU - it needs separate VRs for that):

pk->planes[IACORE].OffsetVolts =  -65;
pk->planes[RING].OffsetVolts = -65;

Same for all other domains (UNCORE, GTSLICE, GTUNSLICE) - right now, you are undervolting GTSLICE and GTUNSLICE domains by -55 mV.

**Comment by trihutama** [comment]
yes i understand, for the core and ring i undervolt both same values for stability. But what confuse me is when i undervolt other domains (UNCORE/GTSLICE/GTUNSLICE) it got overrides by same values from core domains. Just look at report from throttlestop above, it report offset -65 mv for all domains.
I've tried different values and not enable programming VF for other domains (UNCORE, GTSLICE, GTUNSLICE) , but it keep getting overrides by values from core domains.

**Comment by psyq321** [comment]
Can you enable tracing in CONFIGURATION.h and then make a video of the PowerMonkey.efi running with a Smartphone? Ideally it should be one of those capable of high-FPS capture as time lapse, because debug messages display for a very short amount of time.

This way we can see what is it doing exactly and which values are being programmed. Sorry for such a crude method, but running on bare metal leaves very little options except for standard hardware-debug cable + another host with system debugger running, which most people do not have.

**Comment by trihutama** [comment]
i tried to enable tracing, but it gave me this error :
`NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\Hostx86\x64\cl.exe"' : return code '0x2' Stop.`

btw i'm using [Build-PowerMonkey](https://github.com/awyl/Build-PowerMonkey) for build PowerMonkey.efi because i'm not too familiar with edk2. 
do you think this just error report by throttlestop ? because it still on beta for windows 11 ?

**Comment by emc02** [comment]
I have the same i5-8550u in my XPS13 9360 and undervolting CPU and Cache @ -105,5 works well.
Changing the PowerLimits of PL1/PL2 to custom values always results in a emergency mode with 3W Power-Limit and 400MHz Clockspeed (very slow Windows experience).
Setting it to MAX_POWAH and locking everything shows then up as 4095W, but Power Limit is still locked at 23W (tested with Cinebench)
Is there any way to unlock the power limit to 28W or so? (I've watercooled my XPS13 and could get much higher with PL1 ;-) )

**Comment by trihutama** [comment]
@emc02 I think you set wrong configuration, you should always convert W to mW when set custom PL1/PL2. If you want to set it to 28W then it should be set 28000 in your configuration.c . Don't use MAX_POWAH, it's unstable. I try it before and my power limit almost double from 25W to 44W with i5-8250U. It burns my CPU from average 75° to 95° for 10 second before heat throttle.

**Comment by emc02** [comment]
I was setting it to 28000. What Notebook are you using? Maybe Dell is setting harder limits
